src/session/client-certificate.c

```diff
@@ -49,20 +64,34 @@
  * including "0::" prefix and newline.
  * NB: the kernel doesn't allow newlines in cgroup names. */
 static char *
-read_proc_self_cgroup (size_t *out_length)
+read_proc_pid_cgroup (int dirfd, size_t *out_length)
 {
-  FILE *fp = fopen ("/proc/self/cgroup", "r");
-
-  if (fp == NULL)
+  int fd = openat (dirfd, "cgroup", O_RDONLY | O_CLOEXEC);
+  if (fd < 0)
     {
-      warn ("Failed to open /proc/self/cgroup");
+      warn ("Failed to open /proc/pid/cgroup");
       return NULL;
     }
 
   char buffer[1024];
-  *out_length = fread (buffer, 1, sizeof (buffer) - 1, fp);
+  int ret;
+  while (true)
+    {
+      ret = read (fd, buffer, sizeof (buffer));
+      if (ret < 0)
+        {
+          if (errno == EINTR)
+            continue;
+          warn ("Failed to read /proc/pid/cgroup");
+          close (fd);
+          return NULL;
+        }
+      break;
+    }
+
+  *out_length = ret;
   buffer[*out_length] = '\0';
```

Comment:
Danger danger!  You pass `sizeof buffer` to `read()` but then use its return value to write a nul terminator.  If we read exactly `1024` bytes, you'd then set `buffer[1024] = '\0'` which is one past the end of the array.

All in all I think this code is excessively complicated.  The EINTR handling is admirable, but I think that wouldn't happen on `/proc`, and even if it did, you'd also then need to handle partial reads, no?

fwiw, I prefer to write `EINTR` loops like:

```c
do
    ret = read(fd, buffer, sizeof buffer);
while (ret == -1 && errno == EINTR);

// handle ret -1 for non-EINTR cases here...
```

But why not just:

```
ret = read(fd, buffer, sizeof buffer);
if ret < 1 → error
if buffer[ret - 1] != '\n' → error
buffer[ret - 1] = '\0';
return buffer
```

I'm pretty sure this will work:
 - `EINTR` isn't actually gonna happen here
 - if we didn't get the `'\n'` then it means we didn't read the entire cgroup name
 - we should eliminate the trailing `'\n'` anyway, so that's a good spot to put the `'\0'`.

Reply:
Okay, so I read a bit further and we actually want to return the `\n`.  That's fine.  We can add the `\0` after the `\n` but we should still check that `\n` was the last character.

So basically: bring back the `sizeof buffer - 1` as in the `fread()` call, please.

Reply:
>  The EINTR handling is admirable, but I think that wouldn't happen on /proc, and even if it did, you'd also then need to handle partial reads, no?

I think EINTR is less a function of "/proc or not" and more about "I have signal handlers / threads" and such. A read that fails with EINTR did *not* read anything, so it's independent from a partial read. But I suppose you know that... 

I was counting on partial  reads not happening on /proc indeed, as it's all virtual anyway, small, and predictable (and it fails to the right side of security). Indeed I suppose we can drop the EINTR, as the only signal that we expect is the alarm(), and then it's game over anyway.

The code already checks both "too long" and "ends with `\n` a little further down, in the big "if" condition with lots of comments.

Thanks for spotting the off-by-one! Fixed. And I simplified the code to drop the while/EINTR parts.